### PR TITLE
Use fixed pool size, parse all single line matchers in one thread

### PR DESCRIPTION
**Pull request by Jimilian**
So, this is ambitious PR:
- Set language level to Java 7
- Thread pool is constant now - doesn't depend on number of failed job
- All single line patterns are parsed in one thread - reduces CPU load and memory pressure A LOT

I understand that right now this PR is far away from final state. But it already works as expected.
My questions are:
- Can I keep Java 7?
- Can I delete\rework scan(Run build, PrintStream buildLog)?
- Can I rework chain Cause->Indication->Pattern? I don't like current solution `setter`+`getter`, but I didn't want to make significant changes without asking it here.

### CODE REVIEW

**Comment by Jimilian** [comment]
Close and reopen to re-trigger. (Jenkins report check style issues on empty lines)

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/BuildLogFailureReader.java [hunk 1]
```diff
@@ -61,46 +62,23 @@ public BuildLogFailureReader(BuildLogIndication indication) {
...
-     * Scans for indications of a failure cause in a build log. Note: If an exception
-     * occurs during the scanning, information about the exception is appended to
-     * the build log.
-     *
-     * @param build the build to scan for indications.
-     * @param buildLog the log of the build.
-     * @return a FoundIndication if something was found, null if not.
-     */
-    @Override
-    public FoundIndication scan(Run build, PrintStream buildLog) {
```
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/BuildLogFailureReader.java]: Should I keep this method as well?

**Comment by rsandell** [comment]
I'm not sure I understand your motivation for doing this kind of invasive change, to me sacrificing memory footprint for a quicker set of scans is well within reason, hardware is cheap but patient users are not ;)

Anyway I'm going to Jenkins World next week so I won't have time to look at this until I'm back.

**Comment by Jimilian** [comment]
Motivation is pretty simple - I want to be sure that BFA will not kill my Jenkins instance and would be fast enough to provide feedback.
To achieve first goal, I need to make number of threads constant (see picture that can happen in some cases). It allocates a lot of memory (it's 10 seconds snapshot) and run a lot threads ([CPU starvation](https://en.wikipedia.org/wiki/Starvation_(computer_science))).
To achieve second goal, I need to parse as much patterns as possible in one thread. It's quite easy to do, if you have a lot of single line patterns. In this case you can parse all patterns in once.


bfa_rocks

**Comment by Jimilian** [comment]
Close and reopen to re-trigger. (Again statistics tests report wrong date).

**Comment by Jimilian** [comment]
@rsandell I hope that you had recovered already after Jenkins World :)
Do you have a time to take a look into this PR?

I want to mention only one thing that we need to address - `TIMEOUT_FILE`. To keep old behaviour I can multiply it on number of potential causes. Move this value to parameters or just set it in 2 minutes.

**Comment by Jimilian** [comment]
> to me sacrificing memory footprint for a quicker set of scans is well within reason
Just to make it clear. If you have only single line patterns with this PR BFA works much-much faster: 
- there is no need to read file multiple time
- all patterns could be checked at once
- there is no hidden queue (you have only 5 executors and 10 patterns = 5 patterns are waiting free executor)

In case of multiline parser plugin uses constant thread pool size instead of thread pool per build. So, user can specify how much BFA threads he wants: 1, 5 or 10. Depends on machine resources.

**Comment by dcoraboeuf** [comment]
I'm really interested in the outcome of this PR. We have a big volume (> 5000 jobs, logs > 10 Mb, ) and we had to disable BFA because builds were taking too much time to finish.

**Comment by Jimilian** [comment]
@dcoraboeuf , you already can play with this.
It's enough to download hpi artefact from Jenkins: https://jenkins.ci.cloudbees.com/job/plugins/job/build-failure-analyzer-plugin/230/com.sonyericsson.jenkins.plugins.bfa$build-failure-analyzer/artifact/com.sonyericsson.jenkins.plugins.bfa/build-failure-analyzer/1.17.2-SNAPSHOT/build-failure-analyzer-1.17.2-SNAPSHOT.hpi
And install it in "Advanced" panel.

**Comment by dcoraboeuf** [comment]
I'll have to use it in our acceptance environment (I cannot test it in production) but the volume is much lower. And I won't be able to do this for quite some time due to planning constraints :( Thanks all the same, I'll try to follow this.

**Comment by Jimilian** [comment]
@rsandell Any chance to see it in upstream?

**Comment by oleg-nenashev** [comment]
CC @jenkinsci/code-reviewers

**Comment by mmitche** [comment]
I think this is a good PR.  On my instance (very large), we actually can occasionally run out of threads when the BFA is run, which causes cascading issues elsewhere. Specifically, the timeout threads really hurt too.

I'm going to give this is a go.

**Comment by Jimilian** [comment]
@mmitche thx! I was so close to lost hope that even tried to promote this PR: http://www.slideshare.net/jimbrighthead/continuous-integration-on-steroids/59

**Comment by mmitche** [comment]
We have ~30 indications.  There may be other issues, but my instance jumped from ~2k to ~10k threads (and crashed).  this is at least one culprit based on the thread dump.

The issue may be that native resources start to not be collected, and bam bam...done.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java [hunk 2]
```diff
@@ -248,34 +247,75 @@ public void run() {
...
+                List<FoundIndication> foundIndications = new ArrayList<>();
+                for (Indication indication : cause.getIndications()) {
+                    long start = System.currentTimeMillis();
+
+                    Thread.currentThread().setName("BFA-scanner-"
+                            + build.getFullDisplayName() + ": "
+                            + cause.getName() + "-"
+                            + indication.getUserProvidedExpression());
+
+                    foundIndications.add(findIndication(indication, build, buildLog));
```
Comment (mmitche) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: findIndication can return null for the multi-line indications (if they don't match), which adds a null indication and causes an NPE down the line.

File: pom.xml [hunk 3]
```diff
@@ -230,8 +230,8 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-compiler-plugin</artifactId>
                 <configuration>
-                    <source>1.6</source>
-                    <target>1.6</target>
+                    <source>1.7</source>
+                    <target>1.7</target>
```
Comment (oleg-nenashev) [inline pom.xml]: I would go even further:
1) Update parent POM to 2.x latest
2) Update Core requirement to 1.609.x or above
3) Get rid of this section, you will get Java 7 by default

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/BuildLogFailureReader.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/BuildLogFailureReader.java]: I am not sure if deletion of override is required

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java [hunk 4]
```diff
@@ -248,34 +247,75 @@ public void run() {
...
+
+                    if (logger.isLoggable(Level.FINER)) {
+                        logger.log(Level.FINER, "[BFA] [{0}] [{1}] {2}ms", new Object[]{build.getFullDisplayName(),
```
Comment (oleg-nenashev) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: Have you checked that the logger really processes this formatter string with double brackets?

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java [hunk 5]
```diff
@@ -248,34 +247,75 @@ public void run() {
...
+                                String.valueOf(System.currentTimeMillis() - start), });
+                    }
+                }
+
+                foundFailureCauses.add(new FoundFailureCause(cause, foundIndications));
+            }
+
+            return foundFailureCauses;
+        } catch (IOException e) {
+            buildLog.print("[BFA] Exception during parsing file: " + e);
```
Comment (oleg-nenashev) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: Also exception to System log?

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/Indication.java [hunk 6]
```diff
@@ -53,6 +54,8 @@
      */
     private String pattern;
 
+    private FailureCause cause;
```
Comment (oleg-nenashev) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/Indication.java]: Unused field?

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java [hunk 7]
```diff
@@ -58,6 +59,7 @@
     private String pattern;
     private Run build;
     private String matchingString;
+    private FailureCause cause;
```
Comment (oleg-nenashev) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java]: unused field?

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java [hunk 8]
```diff
@@ -58,6 +59,7 @@
     private String pattern;
     private Run build;
```
Comment (mmitche) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java]: Is this supposed to be here?  Doesn't this end up creating a cycle that makes collection difficult?

Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java]: It was added to support Pipeline builds (according `git blame`)

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java (hunk 7 above)
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java]: Done

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java (hunk 4 above)
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: It was here for ages - I didn't change anything :/

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java (hunk 5 above)
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: Should I add it like this
```
if (logger.isLoggable(Level.FINER)) {
            logger.log(Level.FINER, ... );
```
?

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/BuildLogFailureReader.java (hunk 1 above)
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/BuildLogFailureReader.java]: Accident :)

File: pom.xml (hunk 3 above)
Comment (Jimilian) [inline pom.xml]: I hope that I understood that you wanted.

**Comment by rsandell** [comment]
Sorry, this looked to me like a big change or at least touched some critical paths, and I have been trying to carve out a big enough chunk of time in my day to give the review a thorough look and not just a quick glance. But my schedule has just been too full. I will get to it a some point, unless @TWestling wakes up and does it before me ;)

File: pom.xml (hunk 3 above)
Comment (rsandell) [inline pom.xml]: Yes bumping to parent pom 2 would cause too much unrelated changes and make the review even harder/longer ;)

**Comment by mmitche** [comment]
Can confirm this ran in production for quite a while with no issues other than the one I already reviewed.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java [hunk 9]
```diff
@@ -59,7 +58,6 @@
...
     private String matchingString;
-    private FailureCause cause;
```
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java]: @rsandell @oleg-nenashev removing this stuff produces
```
SEVERE: Failed Loading plugin build-failure-analyzer
java.io.IOException: Failed to initialize
	at hudson.ClassicPluginStrategy.load(ClassicPluginStrategy.java:442)
	at hudson.PluginManager$2$1$1.run(PluginManager.java:406)
	at org.jvnet.hudson.reactor.TaskGraphBuilder$TaskImpl.run(TaskGraphBuilder.java:169)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:282)
	at jenkins.model.Jenkins$8.runTask(Jenkins.java:926)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:210)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Unable to read /var/lib/jenkins/build-failure-analyzer.xml
	at hudson.XmlFile.unmarshal(XmlFile.java:165)
	at hudson.Plugin.load(Plugin.java:264)
	at com.sonyericsson.jenkins.plugins.bfa.PluginImpl.start(PluginImpl.java:151)
	at hudson.ClassicPluginStrategy.startPlugin(ClassicPluginStrategy.java:450)
	at hudson.ClassicPluginStrategy.load(ClassicPluginStrategy.java:439)
	... 9 more
Caused by: com.thoughtworks.xstream.converters.ConversionException: null : null
---- Debugging information ----
cause-exception     : java.lang.NullPointerException
cause-message       : null
class               : com.sonyericsson.jenkins.plugins.bfa.PluginImpl
required-type       : com.sonyericsson.jenkins.plugins.bfa.PluginImpl
converter-type      : hudson.util.RobustReflectionConverter
path                : /com.sonyericsson.jenkins.plugins.bfa.PluginImpl
--
java.io.IOException: Unable to load org.jenkinsci.plugins.globalEventsPlugin.GlobalEventsPlugin from groovy-events-listener-plugin
	at hudson.ClassicPluginStrategy.load(ClassicPluginStrategy.java:425)
	at hudson.PluginManager$2$1$1.run(PluginManager.java:406)
	at org.jvnet.hudson.reactor.TaskGraphBuilder$TaskImpl.run(TaskGraphBuilder.java:169)
	at org.jvnet.hudson.reactor.Reactor.runTask(Reactor.java:282)
	at jenkins.model.Jenkins$8.runTask(Jenkins.java:926)
	at org.jvnet.hudson.reactor.Reactor$2.run(Reactor.java:210)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ExceptionInInitializerError
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at hudson.ClassicPluginStrategy.load(ClassicPluginStrategy.java:419)
	... 9 more
```
I restored this line (and from `Indication`) and everything works again. So, what should I do? Restore it? And mark `transient`?

Comment (rsandell) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java]: https://wiki.jenkins-ci.org/display/JENKINS/Hint+on+retaining+backward+compatibility

File: pom.xml (hunk 3 above)
Comment (rsandell) [inline pom.xml]: But you should bump code dependency to 1.625.3 to indicate the Java 7 requirement if you actually need it, otherwise stay on Java 6 as that was the Java version supported by Jenkins up until 1.612

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java (hunk 9 above)
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/FoundIndication.java]: Done.

File: pom.xml (hunk 3 above)
Comment (Jimilian) [inline pom.xml]: Ok, I switched back to Java 6. I wasn't brave enough to fix compilation for all API changes between `1.612` and `1.625.3`.

**Comment by Jimilian** [comment]
@rsandell, @TWestling, what should I do to merge this PR? :)

**Comment by TWestling** [comment]
Sorry, I haven't had the time to look at this thoroughly. I will try to squeeze it in during this week,

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/Indication.java [hunk 10]
```diff
@@ -54,6 +55,12 @@
     private String pattern;
 
     /**
+     * @deprecated, kept for backwards compatibility.
+     */
+    @Deprecated
+    private transient FailureCause cause;
+
```
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/Indication.java]: Unfortunately, it's not possible. I tried - but Jenkins fails to deserialize `config.xml` in this case. I don't remember exactly the error, but it was something about "unknown field".

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java [hunk 11]
```diff
@@ -190,44 +197,38 @@ public static void scan(Run build, PrintStream buildLog) {
...
-        final List<FoundFailureCause> foundFailureCauseList =
-                Collections.synchronizedList(new LinkedList<FoundFailureCause>());
+        final List<FoundFailureCause> foundFailureCauseList = new ArrayList<FoundFailureCause>();
         long start = System.currentTimeMillis();
-        ThreadPoolExecutor executor = (ThreadPoolExecutor)Executors.
-                newFixedThreadPool(PluginImpl.getInstance().getNrOfScanThreads());
+
+        THREAD_POOL_EXECUTOR.setCorePoolSize(PluginImpl.getInstance().getNrOfScanThreads());
+        THREAD_POOL_EXECUTOR.setMaximumPoolSize(PluginImpl.getInstance().getNrOfScanThreads());
+
```
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: So, before user configured number of threads used for single build. Now it's global pool - Jenkins/plugin allows to specify maximum number of threads used for parsing.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java [hunk 12]
```diff
@@ -190,44 +197,38 @@ public static void scan(Run build, PrintStream buildLog) {
...
-                    Thread.currentThread().setName("BFA-scanner-"
-                                                           + build.getFullDisplayName() + ": "
-                                                           + cause.getName());
-                    List<FoundIndication> foundIndications = findIndications(cause, build, buildLog);
-                    if (!foundIndications.isEmpty()) {
-                        FoundFailureCause foundFailureCause = new FoundFailureCause(cause, foundIndications);
-                        foundFailureCauseList.add(foundFailureCause);
-                    }
+                    Thread.currentThread().setName("BFA-scanner-" + build.getFullDisplayName());
+                    foundFailureCauseList.addAll(findIndications(causes, build, buildLog));
```
Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: It could create delay at first time, but after that user can increase number of threads if server has enough capacity or decrease number of regular expressions (or optimize it :) ).

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java (hunk 11 above)
Comment (TWestling) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: Ah, of course, that makes sense.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/Indication.java (hunk 10 above)
Comment (TWestling) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/indication/Indication.java]: This doesn't make sense to me. The field only exists in one of your previous commits of this pull request. So your failure to deserialize should only happen when running an instance where you have ran one of your earlier commits, something that won't happen for anyone else.

File: src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java (hunk 12 above)
Comment (TWestling) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: The way I thought it worked was that you start one thread for all single lines, which is faster than going through the build log several times.

But this one thread is also used for handling everything in the old way, for Causes with at least one non-single-line indication. So in a setup with a lot of non-single-lines, what earlier was done in parallel in a lot of threads, is now done in a serial way, with one thread. 

This doesn't seem to be tweakable in any way for a specific build, only for the server as a whole. If any given log is long and you have lots of MultiLines, it will take a long time, is this correct or am I misreading something?

Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: Yes, you are right. I can add support for this case as well (one task for all pure-single-line regular expressions, and one task per not-only-single-line expression). Should I do it in this PR?

Comment (TWestling) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: Yeah, I think that would make sense, then we aren't sacrificing one use case in favor of another.

Comment (TWestling) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: @Jimilian Is there an ETA for the changes for not-only-single-line expressions? :)

Comment (Jimilian) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: @TWestling Oh, man! One month had passed already! I will try to implement it on next week. But I'm doing it in my spare time...

Comment (TWestling) [inline src/main/java/com/sonyericsson/jenkins/plugins/bfa/BuildFailureScanner.java]: @Jimilian No rush, just thought I'd ping you so that it wasn't forgotten altogether :)

**Comment by Jimilian** [comment]
@rsandell, @TWestling This last commit this PR does:
* One global thread pool for BFA (instead of thread pool per build)
* All single line causes will be parsed together (one shared stream, one thread)
* All non-single-line-only causes  will be parsed in parallel (one stream per regexp, one thread per cause)
* Add several test cases

**Comment by TWestling** [comment]
@Jimilian Sorry about the repeated delays, just as you commited your latest changes, I went on parental leave for a month. I'm back in office today and will take a look at your changes.

**Comment by Jimilian** [comment]
@TWestling, congratulations ;)

**Comment by TWestling** [comment]
@Jimilian Thank you :)

**Comment by mmitche** [comment]
Awesome
